astropy/wcs/src/wcslib_wtbarr_wrap.c

```diff
@@ -0,0 +1,233 @@
+#define NO_IMPORT_ARRAY
+
+#include "astropy_wcs/wcslib_wtbarr_wrap.h"
+
+#include <wcs.h>
+#include <wcsprintf.h>
+#include <tab.h>
+#include <wtbarr.h>
+
+/*
+ It gets to be really tedious to type long docstrings in ANSI C syntax
+ (since multi-line strings literals are not valid).  Therefore, the
+ docstrings are written in doc/docstrings.py, which are then converted
+ by setup.py into docstrings.h, which we include here.
+*/
+#include "astropy_wcs/docstrings.h"
+
+
+/***************************************************************************
+ * PyWtbarr methods                                                        *
+ ***************************************************************************/
+
+static PyObject*
+PyWtbarr_new(PyTypeObject* type, PyObject* args, PyObject* kwds) {
+  PyWtbarr* self;
+  self = (PyWtbarr*)type->tp_alloc(type, 0);
+  return (PyObject*)self;
+}
+
+
+static int
+PyWtbarr_traverse(PyWtbarr* self, visitproc visit, void *arg) {
+  Py_VISIT(self->owner);
+  return 0;
+}
+
+
+static int
+PyWtbarr_clear(PyWtbarr* self) {
+  Py_CLEAR(self->owner);
+  return 0;
+}
+
+
+static void PyWtbarr_dealloc(PyWtbarr* self) {
+  PyWtbarr_clear(self);
+  Py_TYPE(self)->tp_free((PyObject*)self);
+}
+
+
+PyWtbarr* PyWtbarr_cnew(PyObject* wcsprm, struct wtbarr* x) {
+  PyWtbarr* self;
+  self = (PyWtbarr*)(&PyWtbarrType)->tp_alloc(&PyWtbarrType, 0);
+  if (self == NULL) return NULL;
+  self->x = x;
+  Py_INCREF(wcsprm);
+  self->owner = wcsprm;
+  return self;
+}
+
+
+static void wtbarrprt(const struct wtbarr *wtb) {
+  int i, nd, ndim;
+
+  if (wtb == 0x0) return;
+
+  wcsprintf("     i: %d\n", wtb->i);
+  wcsprintf("     m: %d\n", wtb->m);
+  wcsprintf("  kind: %c\n", wtb->kind);
+  wcsprintf("extnam: %s\n", wtb->extnam);
+  wcsprintf("extver: %d\n", wtb->extver);
+  wcsprintf("extlev: %d\n", wtb->extlev);
+  wcsprintf(" ttype: %s\n", wtb->ttype);
+  wcsprintf("   row: %ld\n", wtb->row);
+  wcsprintf("  ndim: %d\n", wtb->ndim);
+  wcsprintf("dimlen: %#lx\n", (unsigned long int)wtb->dimlen);
```

Comment:
I'm not completely sure, but I think pointer addresses should be printed using `%p` and the pointer needs to be cast to `void *`.

Not sure if this could lead to real problems but at least on windows long is only 32bits (one needs to use long long for 64bit).

Reply:
to the best of my knowledge (which is kind of rusty) pointers are `DWORD` on Windows (i.e., `unsigned long`) but I may be completely wrong. Also, I essentially tried to reproduce the same formatting the WCSLIB is using, see https://github.com/astropy/astropy/blob/918abcd480acdf824c5e5d856daf0573811d9b3c/cextern/wcslib/C/wcsprintf.h#L136-L141

Reply:
My knowledge is also a bit lacking but I think windows uses 48bits for pointers and the type is `unsigned long long`.

But wouldn't it be best to use the `%p`placeholder and `void *`-casting - or does this have other issues? That's what [CPython also uses for `id` and the `repr` of objects without `__repr__`](https://github.com/python/cpython/blob/v3.8.0/Objects/object.c#L518-L520)